Source/Scene/Model/MetadataPipelineStage.js

```diff
@@ -62,62 +83,227 @@ MetadataPipelineStage.process = function (
-
-  for (const metadataType of metadataTypes) {
-    const structName = `${metadataType}MetadataClass`;
-    const structIdVs = `${structName}VS`;
-    const structIdFs = `${structName}FS`;
-    // Declare the struct in both vertex and fragment shaders
-    shaderBuilder.addStruct(structIdVs, structName, ShaderDestination.VERTEX);
-    shaderBuilder.addStruct(structIdFs, structName, ShaderDestination.FRAGMENT);
-
-    // Add fields
-    for (let j = 0; j < classFields.length; j++) {
-      const shaderName = classFields[j].shaderName;
-      shaderBuilder.addStructField(structIdVs, metadataType, shaderName);
-      shaderBuilder.addStructField(structIdFs, metadataType, shaderName);
+function declareMetadataTypeStructs(shaderBuilder, propertyInfos) {
+  const classTypes = new Set();
+  const statisticsTypes = new Set();
+  const enumLengths = new Set();
+
+  for (let i = 0; i < propertyInfos.length; i++) {
+    const { type, enumType, glslType, propertyStatistics } = propertyInfos[i];
+    classTypes.add(glslType);
+    if (!defined(propertyStatistics)) {
+      continue;
+    }
+    if (type === MetadataType.ENUM) {
+      enumLengths.add(enumType.values.length);
+    } else {
+      statisticsTypes.add(glslType);
+    }
+  }
+
+  const classFields = MetadataPipelineStage.METADATA_CLASS_FIELDS;
+  for (const metadataType of classTypes) {
+    const classStructName = `${metadataType}MetadataClass`;
+    declareTypeStruct(classStructName, metadataType, classFields);
+  }
+
+  const statisticsFields = MetadataPipelineStage.METADATA_STATISTICS_FIELDS;
+  for (const metadataType of statisticsTypes) {
+    const statisticsStructName = `${metadataType}MetadataStatistics`;
+    declareTypeStruct(statisticsStructName, metadataType, statisticsFields);
+  }
+
+  for (const length of enumLengths) {
+    if (length < 1) {
+      continue;
+    }
+    const structName = `enum${length}MetadataStatistics`;
+    shaderBuilder.addStruct(structName, structName, ShaderDestination.BOTH);
+
+    const shaderType = "int";
+    const shaderName = `occurrences[${length}]`;
+    shaderBuilder.addStructField(structName, shaderType, shaderName);
+  }
+
+  function declareTypeStruct(structName, type, fields) {
+    shaderBuilder.addStruct(structName, structName, ShaderDestination.BOTH);
+
+    for (let i = 0; i < fields.length; i++) {
+      const { shaderName } = fields[i];
+      const shaderType =
+        fields[i].type === "float" ? convertToFloatComponents(type) : type;
+      shaderBuilder.addStructField(structName, shaderType, shaderName);
     }
   }
 }
 
+/**
+ * For a type with integer components, find a corresponding float-component type
+ * @param {String} type The name of a GLSL type with integer components
+ * @returns {String}
+ * @private
+ */
+function convertToFloatComponents(type) {
+  const conversions = {
+    int: "float",
+    ivec2: "vec2",
+    ivec3: "vec3",
+    ivec4: "vec4",
+  };
```

Comment:
Since this object doesn't change, it could be pulled out of this function as a constant